### PR TITLE
[status=preview] Allow grid to use grpc custom encodings for types

### LINKED ISSUE #80: Add support for custom message encoding/decoding.
Currently all actor to actor messages are encoded via Gob.   While convenient and pretty fast, gob isn't as efficient as protobuf or flatbuffers.

### PR DESCRIPTION
**Pull request by epsniff**
Closes #80 

TODOs
- [ ] review this with @mdmarek.  
- [ ] write protobuf codecs for all the internal messages. 
- [ ] write an example of using flatbuffers
- [ ] get all the tests passing.  Currently some deadlock.  Most of these have been where we drop errors without logging or returning them.  So now we're discovering those issues which is good :).

NOTE: This design was driven a lot by the way we allow callers to send/receive interface{} types into grid.

### CODE REVIEW

File: grid.v3/msg.go [hunk 1]
```diff
@@ -43,6 +44,35 @@ type ActorStart struct {
 	Data []byte
 }
 
+type ackmsg struct {
```
Comment (epsniff) [inline grid.v3/msg.go]: this msg is both the message an it's codec.  It also doesn't unmashal anything as it can just return the global Ack message.

**Comment by araddon** [comment]
cool.

File: grid.v3/client.go [hunk 2]
```diff
@@ -242,26 +241,27 @@ func (c *Client) RequestC(ctx context.Context, receiver string, msg interface{})
...
-		return nil, err
+	if len(res.Data) == 0 {
+		return nil, ErrNilResponse
 	}
-	if n != len(res.Data) {
-		return nil, io.ErrUnexpectedEOF
+
+	if res.CodecName == "" {
+		c.logf("%v: error no codec provided in response", c.cfg.Namespace)
+		return nil, ErrUnRegisteredMsgType
```
Comment (epsniff) [inline grid.v3/client.go]: This should gets it's own Error, and not use UnRegistered

File: grid.v3/client_test.go [hunk 3]
```diff
@@ -63,6 +66,14 @@ func (a *echoActor) Act(c context.Context) {
 	}
 }
 
+type EchoMsg struct {
+	Msg string
+}
+
```
Comment (epsniff) [inline grid.v3/client_test.go]: `GobCodecRegister` is a convenience method that generates a GOB codec for a type and registers it.

**Pull request merged**